docs/posts/flink-quickstart/index.qmd

```diff
@@ -0,0 +1,259 @@
+If you don't intend to try [remote execution](#remote-execution), you can start only the Kafka-related services with `docker compose up kafka init-kafka data-generator`.
+:::
+
+After a few seconds, you should see messages indicating your Kafka environment is ready:
+
+```bash
+ibis-flink-example-init-kafka-1      | Successfully created the following topics:
+ibis-flink-example-init-kafka-1      | payment_msg
+ibis-flink-example-init-kafka-1      | sink
+ibis-flink-example-init-kafka-1 exited with code 0
+ibis-flink-example-data-generator-1  | Connected to Kafka
+ibis-flink-example-data-generator-1  | Producing 20000 records to Kafka topic payment_msg
+```
+
+This example uses mock payments data. The `payment_msg` Kafka topic contains messages in the following format:
+
+```json
+{
+    "createTime": "2023-09-20 22:19:02.224",
+    "orderId": 1695248388,
+    "payAmount": 88694.71922270155,
+    "payPlatform": 0,
+    "provinceId": 6
+}
+```
+
+In a separate terminal, we can explore what these messages look like:
+
+```{python}
+from itertools import islice
+
+from kafka import KafkaConsumer
+
+consumer = KafkaConsumer("payment_msg")
+for msg in islice(consumer, 3):
+    print(msg)
+```
+
+## Running the tutorial
+
+This tutorial uses the Flink backend for Ibis to process the aforementioned payment messages. You can choose to either [run it locally](#local-execution) or [submit a job to an already-running Flink cluster](#remote-execution).
+
+### Local execution
+
+The simpler option is to run the example using the Flink mini cluster.
+
+#### Create a table environment
+
+The [table environment](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/table_environment/) serves as the main entry point for interacting with the Flink runtime. The `flink` backend does not create `TableEnvironment` objects; you must create a `TableEnvironment` and pass that to [`ibis.flink.connect`](../../backends/flink.qmd#ibis.flink.connect).
+
+```{python}
+import ibis
+from pyflink.table import EnvironmentSettings, TableEnvironment
+
+env_settings = EnvironmentSettings.in_streaming_mode()
+table_env = TableEnvironment.create(env_settings)
+# write all the data to one file
+table_env.get_config().set("parallelism.default", "1")
+
+con = ibis.flink.connect(table_env)
+```
+
+Flink’s streaming connectors aren't part of the binary distribution. Link the [Kafka connector](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/connectors/table/kafka/) for cluster execution by adding the JAR file from the cloned repository.
+
+```{python}
+#| echo: false
+#| output: false
+
+import urllib.request
+
+urllib.request.urlretrieve(
+    "https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka/3.0.2-1.18/flink-sql-connector-kafka-3.0.2-1.18.jar",
+    filename="flink-sql-connector-kafka-3.0.2-1.18.jar",
+)
+```
+
+```{python}
+#| output: false
+
+con._exec_sql("ADD JAR 'flink-sql-connector-kafka-3.0.2-1.18.jar'")
```

Comment:
this looks less than ideal calling a private method? could this be `raw_sql()`?

Reply:
looks like it's not implemented right now

Reply:
Hmm... I was just looking at it; I think we could pretty easily implement `raw_sql()`? Let me take a look tonight...

Reply:
https://github.com/ibis-project/ibis/pull/8104